crates/cadenza-eval/src/interner.rs

```diff
@@ -33,6 +66,423 @@ impl fmt::Display for InternedId {
+            }
+
+            fn with_mut<R>(
+                f: impl FnOnce(&mut $crate::interner::InternData<$value_type>) -> R,
+            ) -> R {
+                STORAGE.with(|data| f(&mut data.borrow_mut()))
+            }
+        }
+    };
+}
+
+// =============================================================================
+// Generic InternId with Type Parameter
+// =============================================================================
+
+/// A generic interned ID that carries type information.
+///
+/// `InternId<T>` represents an index into a storage of type `T`.
+/// It implements `Copy`, `Eq`, `Hash` with zero-cost comparison.
+pub struct InternId<T> {
+    index: u32,
+    _marker: PhantomData<fn() -> T>,
+}
+
+// Manual implementations to avoid requiring T: Copy/Clone
+impl<T> Clone for InternId<T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<T> Copy for InternId<T> {}
+
+impl<T> PartialEq for InternId<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.index == other.index
+    }
+}
+
+impl<T> Eq for InternId<T> {}
+
+impl<T> std::hash::Hash for InternId<T> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.index.hash(state);
+    }
+}
+
+impl<T> InternId<T> {
+    /// Creates a new InternId with the given index.
+    #[inline]
+    fn new(index: u32) -> Self {
+        Self {
+            index,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns the raw index of this interned ID.
+    #[inline]
+    pub fn index(self) -> u32 {
+        self.index
+    }
+
+    /// Resolves this ID using the given storage.
+    #[inline]
+    pub fn resolve<S: Storage<T>>(self, storage: &S) -> Option<T>
+    where
+        T: Clone,
+    {
+        storage.with(|data| data.get(self).cloned())
+    }
+
+    /// Resolves this ID using static storage.
+    ///
+    /// This allows resolving the ID anywhere in the program where the
+    /// static storage type `S` is accessible.
+    #[inline]
+    pub fn resolve_static<S: StaticStorage<T>>(self) -> Option<T>
+    where
+        T: Clone,
+    {
+        S::with(|data| data.get(self).cloned())
+    }
+}
+
+impl<T> fmt::Debug for InternId<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "InternId({})", self.index)
+    }
+}
+
+impl<T> fmt::Display for InternId<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.index)
+    }
+}
+
+// =============================================================================
+// Intern: The main interning interface
+// =============================================================================
+
+/// A generic interner that can intern values with optional transformation.
+///
+/// `Intern<T, S>` interns values of type `T` into storage `S`.
+/// It supports deduplication based on a key type `K`.
+#[derive(Debug)]
+pub struct Intern<T, K = T, S = LocalStorage<T>>
```

Comment:
I don't want two structs. Just a single one should be sufficient. We can just call it `Interned<Storage>`. It should have a `fn from(v: &str) -> Self` method. It should implement Deref, which I guess means that the storage needs to have a static lifetime, which is fine. The storage trait should have an `fn insert(v: &str) -> Self::Index` and a `fn resolve(index: Self::Index) -> &'static Self::Value`. For the local storage ones we'd want to make sure the ZST doesn't implement Sync or Send so the intern value can't be sent to a different thread.